crypto/bio/bio_lib.c

```diff
@@ -848,29 +857,20 @@ int BIO_connect_retry(BIO *bio, int timeout)
         return 0;
     }
 
+    if (polling_milliseconds < 0)
+        polling_milliseconds = 100;
     if (!blocking)
         BIO_set_nbio(bio, 1);
```

Comment:
I'd suggest mak this BIO_set_nbio(bio, !blocking);
You can't be sure what the internal state of this bio is?

Reply:
I think in case of blocking mode this should just call BIO_do_connect
there is no need for any retires in that case.
What exactly is the motivation for the non-blocking mode here,
put a time-out how long the connect may take?

Reply:
> I'd suggest mak this BIO_set_nbio(bio, !blocking);
> You can't be sure what the internal state of this bio is?

Done.

Reply:
> I think in case of blocking mode this should just call BIO_do_connect
> there is no need for any retires in that case.
> What exactly is the motivation for the non-blocking mode here,
> put a time-out how long the connect may take?

The retry is still needed in the non-blocking case, as meanwhile made clear in the revised code and its comments:
Even for blocking connects non-local connection attempts may return a premature error in case there a system-internal timeout occurs (e.g., when the intended host is currently not reachable) and in case a domain name resolution fails temporarily.
I've confirmed these situation in my yesterday's experiments.

Reply:
> Done.

but not pushed :-)

> The retry is still needed 

So you re-try the connect even if it is a hard failure.
Now I understand at least what the code should do...

Reply:
Oops, that commit went wrong. Now it's in and pushed :)